File: libgdx_galaxy_invaders/core/src/com/nopalsoft/invaders/game_objects/AlienShip.java
package com.nopalsoft.invaders.game_objects;

public class AlienShip extends DynamicGameObject {

    public static final float RADIUS = 1.5f;

    public static final float DRAW_WIDTH = 3.5f;
    public static final float DRAW_HEIGHT = 3.5f;

    public static final int MOVE_SIDES = 0;
    public static final int MOVE_DOWN = 2;
    public static final int EXPLODING = 3;
    public static final float SPEED = 4f;
    public static final float SPEED_DOWN = -3.5f;

    public static final float HORIZONTAL_MOVEMENT_RANGE = 6.7f;
    public static final float VERTICAL_MOVEMENT_RANGE = 1.2f;
    public static final float EXPLOSION_DURATION = 0.05f * 19;

    public final int POINTS_PER_HIT = 10;

    public int remainingLives;
    public int score;
    public float stateTime;
    public int state;
    float movedDistance;
    float speedIncrease;

    public AlienShip(int vida, float speedIncrease, float x, float y) {
        super(x, y, RADIUS);
        stateTime = 0;
        state = MOVE_SIDES;
        velocity.set(SPEED, SPEED_DOWN);
        movedDistance = 0;
        score = POINTS_PER_HIT;
        remainingLives = vida;
        this.speedIncrease = 1 + speedIncrease;
    }

    public void update(float deltaTime) {
        if (state != EXPLODING) {
            switch (state) {
                case MOVE_SIDES:
                    position.x += velocity.x * deltaTime * speedIncrease;
                    movedDistance += Math.abs(velocity.x * deltaTime) * speedIncrease;
                    if (movedDistance > HORIZONTAL_MOVEMENT_RANGE) {
                        state = MOVE_DOWN;
                        velocity.x *= -1;
                        movedDistance = 0;
                    }
                    break;
                case MOVE_DOWN:
                    position.y += velocity.y * deltaTime * speedIncrease;
                    movedDistance += Math.abs(velocity.x * deltaTime) * speedIncrease;
                    if (movedDistance > VERTICAL_MOVEMENT_RANGE) {
                        state = MOVE_SIDES;
                        movedDistance = 0;
                    }
                    break;
            }
        }

        boundsCircle.x = position.x;
        boundsCircle.y = position.y;
        stateTime += deltaTime;
    }

    /**
     * Calling this method is bullet power 1
     */
    public void beingHit() {
        remainingLives--;
        if (remainingLives <= 0) {
            state = EXPLODING;
            velocity.add(0, 0);
            stateTime = 0;
        }
    }
}

File: libgdx_galaxy_invaders/core/src/com/nopalsoft/invaders/game_objects/AlienShip.kt
package com.nopalsoft.invaders.game_objects

import kotlin.math.abs

class AlienShip(vida: Int, speedIncrease: Float, x: Float, y: Float) : DynamicGameObject(x, y, RADIUS) {
    val POINTS_PER_HIT: Int = 10

    @JvmField
    var remainingLives: Int

    @JvmField
    var score: Int

    @JvmField
    var stateTime: Float = 0f

    @JvmField
    var state: Int
    var movedDistance: Float
    var speedIncrease: Float

    init {
        state = MOVE_SIDES
        velocity[SPEED] = SPEED_DOWN
        movedDistance = 0f
        score = POINTS_PER_HIT
        remainingLives = vida
        this.speedIncrease = 1 + speedIncrease
    }

    fun update(deltaTime: Float) {
        if (state != EXPLODING) {
            when (state) {
                MOVE_SIDES -> {
                    position.x += velocity.x * deltaTime * speedIncrease
                    movedDistance += abs((velocity.x * deltaTime)) * speedIncrease
                    if (movedDistance > HORIZONTAL_MOVEMENT_RANGE) {
                        state = MOVE_DOWN
                        velocity.x *= -1f
                        movedDistance = 0f
                    }
                }

                MOVE_DOWN -> {
                    position.y += velocity.y * deltaTime * speedIncrease
                    movedDistance += abs((velocity.x * deltaTime)) * speedIncrease
                    if (movedDistance > VERTICAL_MOVEMENT_RANGE) {
                        state = MOVE_SIDES
                        movedDistance = 0f
                    }
                }
            }
        }

        boundsCircle!!.x = position.x
        boundsCircle!!.y = position.y
        stateTime += deltaTime
    }

    /**
     * Calling this method is bullet power 1
     */
    fun beingHit() {
        remainingLives--
        if (remainingLives <= 0) {
            state = EXPLODING
            velocity.add(0f, 0f)
            stateTime = 0f
        }
    }

    companion object {
        const val RADIUS: Float = 1.5f

        const val DRAW_WIDTH: Float = 3.5f
        const val DRAW_HEIGHT: Float = 3.5f

        const val MOVE_SIDES: Int = 0
        const val MOVE_DOWN: Int = 2
        const val EXPLODING: Int = 3
        const val SPEED: Float = 4f
        const val SPEED_DOWN: Float = -3.5f

        const val HORIZONTAL_MOVEMENT_RANGE: Float = 6.7f
        const val VERTICAL_MOVEMENT_RANGE: Float = 1.2f
        const val EXPLOSION_DURATION: Float = 0.05f * 19
    }
}
